### PR TITLE
Fix ros2 bugs

### LINKED ISSUE #105: ROS2: High CPU load of python scripts
Getting the following error:
`[ERROR] [ensenso_camera_node-10]: process has died [pid 17673, exit code -9`
when running the `ensenso_camera_node` with a virtual camera and virtual objects. It crashes after around ~300-350 seconds.

### LINKED ISSUE #107: ROS2: High RAM usage of ensenso_camera_node
When running `ensenso_camera_node` and the `image_stream` script the camera node's memory usage increases over time. This does not happen with e.g. the `request_data`script, which in comparison to the `image_stream` script does not include the data in the action response and does not publish any data.

### PR DESCRIPTION
**Pull request by benthie**
Fixes #105
Fixes #107

### CODE REVIEW

File: ensenso_camera_test/src/ensenso_camera_test/telecentric_projection.py [hunk 1]
```diff
@@ -62,9 +62,9 @@ def setUp(self):
         # Test the subscribed cloud
         self.send_goal_with_publishing_point_cloud()
 
-        # Explicitly spin the node so that the subscription callback will be executed, because the test cases do not
-        # send any further actions, which would implicitly spin the node
-        ros2py.spin(self.node)
+        # # Explicitly spin the node so that the subscription callback will be executed, because the test cases do not
+        # # send any further actions, which would implicitly spin the node
+        # ros2py.spin(self.node)
```
Comment (saierd) [inline ensenso_camera_test/src/ensenso_camera_test/telecentric_projection.py:67]: Should be removed completely from both tests.

File: ensenso_camera2/include/ensenso_camera/queued_action_server.h [hunk 2]
```diff
@@ -49,11 +49,16 @@ class QueuedActionServer
                      bool autoStart = false)
     : nh(nh), name(name), callback(callback)
   {
+    // Reduce server's result caching time from 15min by defalt to 10s. The default settings caused the stereo_camera
+    // node to have a huge RAM usage when e.g. the request_data result included images.
+    auto options = rcl_action_server_get_default_options();
+    options.result_timeout.nanoseconds = 1e9 * 10;
```
Comment (saierd) [inline ensenso_camera2/include/ensenso_camera/queued_action_server.h:55]: 10s can easily be 100 frames, which takes a lot of RAM.

https://design.ros2.org/articles/actions.html#result-caching mentions setting the timeout to 0. Does that work? I don't really understand the new API, but does that mean that the client has to start waiting for the result before it gets submitted? If so, that sounds like a race condition...

**Review by saierd — CHANGES_REQUESTED** [review]

File: ensenso_camera2/include/ensenso_camera/queued_action_server.h [hunk 3]
```diff
@@ -49,11 +49,16 @@ class QueuedActionServer
                      bool autoStart = false)
     : nh(nh), name(name), callback(callback)
   {
+    // Reduce server's result caching time from 15min by defalt to 10s. The default settings caused the stereo_camera
```
Comment (saierd) [inline ensenso_camera2/include/ensenso_camera/queued_action_server.h:52]:
```suggestion
    // Reduce server's result caching time from 15min by default to 10s. The default settings caused the stereo_camera
```

File: ensenso_camera2/include/ensenso_camera/queued_action_server.h (hunk 2 above)
Comment (benthie) [reply ensenso_camera2/include/ensenso_camera/queued_action_server.h]: Timeout 0 works.

---

In the Python compatibility layer code you can see how the new API works. It first sends the goal (which optionally takes a feedback callback) and then registers a done-callback in which the client receives a goal handle (future) that contains the response (response and result are separated in ROS2).
https://github.com/ensenso/ros_driver/blob/3cf626be115526715f6eb8f97a8e08ac9c5cd7ff/ensenso_camera/src/ensenso_camera/ros2.py#L74-L77

In the response callback another callback is then registered to the actual result.
https://github.com/ensenso/ros_driver/blob/3cf626be115526715f6eb8f97a8e08ac9c5cd7ff/ensenso_camera/src/ensenso_camera/ros2.py#L85-L93

In the final result callback a threading event is set to indicate that the result has been received.
https://github.com/ensenso/ros_driver/blob/3cf626be115526715f6eb8f97a8e08ac9c5cd7ff/ensenso_camera/src/ensenso_camera/ros2.py#L95-L97

The `wait_for_response` method waits until the threading event has been set. If the method is called and the event has been set already, it does not block according to its API.
https://github.com/ensenso/ros_driver/blob/3cf626be115526715f6eb8f97a8e08ac9c5cd7ff/ensenso_camera/src/ensenso_camera/ros2.py#L79-L83

As far as I understand the only possible spot for a race condition is between L76 and L77, but I think that the ROS2 developers took that into consideration.

**Review by saierd — APPROVED** [review]

**Review by saierd — APPROVED** [review]

**Pull request merged**